sympy/matrices/matrixbase.py

```diff
@@ -1261,78 +1286,74 @@ def wilkinson(kls, n, **kwargs):
         n = as_int(n)
         return klass._eval_wilkinson(n)
 
+    # The RepMatrix subclass uses more efficient sparse implementations of
+    # _eval_iter_values and other things.
+
+    def _eval_iter_values(self):
+        return (i for i in self if i is not S.Zero)
+
+    def _eval_values(self):
+        return list(self.iter_values())
+
+    def _eval_iter_items(self):
+        for i in range(self.rows):
+            for j in range(self.cols):
+                if self[i, j]:
+                    yield (i, j), self[i, j]
+
     def _eval_atoms(self, *types):
         result = set()
         for i in self:
             result.update(i.atoms(*types))
         return result
 
     def _eval_free_symbols(self):
-        return set().union(*(i.free_symbols for i in self if i))
+        return set().union(*(i.free_symbols for i in set(self.values())))
```

Comment:
Is it not possible to do?

```suggestion
        return set().union(*(i.free_symbols for i in self.values()))
```

Reply:
These are equivalent in terms of output but not in terms of runtime. The question is how likely it is that some of the elements of the matrix will be the exact same expression. If every expression is distinct then calling `set(self.values())` is wasteful because constructing the set is unnecessary. If the number of unique expressions is much smaller than the total number of nonzero elements then it is faster to use `set(self.values())` because it avoids calling free_symbols multiple times.

Here is an example of a matrix with 1000000 elements but only 100 unique elements:
```python
In [21]: M = randMatrix(1000)

In [22]: len(M.values())
Out[22]: 989957

In [23]: len(set(M.values()))
Out[23]: 99
```
We can time the two approaches:
```python
In [24]: %time set().union(*(e.free_symbols for e in set(M.values())))
CPU times: user 1.58 s, sys: 0 ns, total: 1.58 s
Wall time: 1.58 s
Out[24]: set()

In [25]: %time set().union(*(e.free_symbols for e in M.values()))
CPU times: user 4.55 s, sys: 149 ms, total: 4.7 s
Wall time: 4.7 s
Out[25]: set()
```
The difference would be more noticeable if the expressions were more complicated and actually had many free symbols etc.

Here is an example where we have the same more complicated expression appearing many times:
```python
In [26]: e = (x + y + z + t)

In [27]: for _ in range(5): e = e**2 + e

In [28]: M = e*eye(100)

In [29]: %time set().union(*(e.free_symbols for e in set(M.values())))
CPU times: user 1.04 ms, sys: 0 ns, total: 1.04 ms
Wall time: 1.06 ms
Out[29]: {t, x, y, z}

In [30]: %time set().union(*(e.free_symbols for e in M.values()))
CPU times: user 55.6 ms, sys: 0 ns, total: 55.6 ms
Wall time: 54 ms
Out[30]: {t, x, y, z}
```
This is because it is faster to identify unique expressions with set() than it is to call free_symbols repeatedly for multiple occurrences of the same expression.

Here is an example where calling `set(self.values())` does not help at all:
```python
In [31]: M = Matrix(1000, 1000, lambda i, j: 1000000*i + j)

In [32]: %time set().union(*(e.free_symbols for e in set(M.values())))
CPU times: user 7 s, sys: 151 ms, total: 7.15 s
Wall time: 7.15 s
Out[32]: set()

In [33]: %time set().union(*(e.free_symbols for e in M.values()))
CPU times: user 7.41 s, sys: 172 ms, total: 7.58 s
Wall time: 7.58 s
Out[33]: set()
```
If you run repeated timings of this then sometimes one method seems faster and sometimes the other. Basically there is no difference. That is because creating a set in `set(self.values())` is cheap compared to everything else that is happening. Sometimes it offers a win by reducing the number of times that `free_symbols` is called. Other times it does not reduce the `free_symbols` calls and is just overhead but it is still a small overhead in that case.

Reply:
Okay, however, I rather think that the definition of ‘values’ seems not very useful.
If it just outputs the list of nonzero elements, without position, it does not seem like more useful than ‘set(m.values())’

(That has backward compatibility issue, however, but I think that some existing method confuses me as much as I thought it would be same in runtime performance as set(m.values()))

Reply:
What about other things like ‘has’, can they rather be improved by set(values)?

Reply:
The implementation of has is in this PR is this:
```python
    def _eval_has(self, *patterns):
        return any(a.has(*patterns) for a in self.iter_values())
```
Here calling `set(self.iter_values())` would be very different because it would force a full iteration over all values whereas `any` can shortcut a full iteration. For example:
```python
In [43]: M = x*randMatrix(1000)

In [44]: %time any(e.has(x) for x in M.iter_values())
CPU times: user 32 ms, sys: 0 ns, total: 32 ms
Wall time: 30.9 ms
Out[44]: True

In [45]: %time any(e.has(x) for x in set(M.iter_values()))
CPU times: user 224 ms, sys: 3.8 ms, total: 228 ms
Wall time: 225 ms
Out[45]: True
```
I'm not sure why the first case is so slow. I think it might be because of caching...

There is an advantage in using `iter_values` if we might shortcut full iteration which we can for `has` but nor for `free_symbols`.

Reply:
> I'm not sure why the first case is so slow. I think it might be because of caching...

The code that I timed was incorrect. Here is the correct version:
```python
In [53]: M = x*randMatrix(1000)

In [54]: %time any(e.has(x) for e in M.iter_values())
CPU times: user 117 µs, sys: 0 ns, total: 117 µs
Wall time: 130 µs
Out[54]: True

In [55]: %time any(e.has(x) for e in set(M.iter_values()))
CPU times: user 228 ms, sys: 3.58 ms, total: 232 ms
Wall time: 230 ms
Out[55]: True
```
Using `set()` is 200x slower in this case.